exec/java-exec/src/main/java/org/apache/drill/exec/rpc/control/WorkEventBus.java

```diff
@@ -74,83 +70,42 @@ public void statusUpdate(final FragmentStatus status) {
-        + QueryIdHelper.getQueryIdentifier(handle));
+  public FragmentManager getFragmentManager(final FragmentHandle handle) {
+    return managers.get(handle);
   }
 
   /**
-   * Removes fragment manager (for the corresponding the handle) from the work event bus. This method can be called
-   * multiple times. The manager will be removed only once (the first call).
-   * @param handle the handle to the fragment
-   */
-  public void removeFragmentManager(final FragmentHandle handle) {
-    if (logger.isDebugEnabled()) {
-      logger.debug("Removing fragment manager: {}", QueryIdHelper.getQueryIdentifier(handle));
-    }
-
-    synchronized (this) {
-      final FragmentManager manager = managers.get(handle);
-      if (manager != null) {
-        recentlyFinishedFragments.put(handle, 1);
-        managers.remove(handle);
-      } else {
-        logger.warn("Fragment {} not found in the work bus.", QueryIdHelper.getQueryIdentifier(handle));
-      }
-    }
-  }
-
-  /**
-   * Cancels and removes fragment manager (for the corresponding the handle) from the work event bus, Currently, used
-   * for fragments waiting on data (root and intermediate).
+   * Optionally cancels and removes fragment manager (for the corresponding the handle) from the work event bus. Currently, used
+   * for fragments waiting on data (root and intermediate). This method can be called multiple times. The manager will be removed
+   * only once (the first call).
    * @param handle the handle to the fragment
+   * @param cancel
    * @return if the fragment was found and removed from the event bus
    */
-  public boolean cancelAndRemoveFragmentManagerIfExists(final FragmentHandle handle) {
-    if (logger.isDebugEnabled()) {
-      logger.debug("Cancelling and removing fragment manager: {}", QueryIdHelper.getQueryIdentifier(handle));
-    }
-
-    synchronized (this) {
-      final FragmentManager manager = managers.get(handle);
-      if (manager == null) {
-        return false;
+  public boolean removeFragmentManager(final FragmentHandle handle, final boolean cancel) {
+    final FragmentManager manager = managers.remove(handle);
+    if (manager != null) {
+      assert !manager.isCancelled() : String.format("Fragment {} manager {} is already cancelled.", QueryIdHelper.getQueryIdentifier(handle), manager);
+      if (cancel) {
+        manager.cancel();
+      }
+      if (logger.isDebugEnabled()) {
+        logger.debug("{} fragment {} manager {} from the work bus.", cancel ? "Cancel and removed" : "Removed",
+            QueryIdHelper.getQueryIdentifier(handle), manager);
       }
-
-      manager.cancel();
-      recentlyFinishedFragments.put(handle, 1);
-      managers.remove(handle);
       return true;
+    } else if (logger.isWarnEnabled()) {
+      logger.warn("Fragment {} manager is not found in the work bus.", QueryIdHelper.getQueryIdentifier(handle));
     }
+    return false;
```

Comment:
Seems like this is redundant. Doesn't seem like we ever return false, and the callers never seem to check.

Reply:
ControlMessageHandler checks the result on line 196.